crates/turborepo-lsp/src/lib.rs

```diff
@@ -0,0 +1,853 @@
+            .workspaces
+            .into_iter()
+            .flat_map(|wd| PackageJson::load(&wd.package_json).ok()) // if we can't parse a package.json, then we can't infer its tasks
+            .collect::<Vec<_>>();
+
+        let tasks = package_jsons
+            .iter()
+            .flat_map(|p| p.scripts.keys())
+            .unique()
+            .map(|s| CompletionItem {
+                label: s.to_owned(),
+                kind: Some(CompletionItemKind::FIELD),
+                ..Default::default()
+            });
+
+        let keys = package_jsons
+            .iter()
+            .flat_map(|p| p.scripts.keys().map(move |k| (p.name.clone(), k)))
+            .map(|(package, s)| CompletionItem {
+                label: format!("{}#{}", package.unwrap_or_default(), s),
+                kind: Some(CompletionItemKind::FIELD),
+                ..Default::default()
+            });
+
+        Ok(Some(CompletionResponse::Array(keys.chain(tasks).collect())))
+    }
+}
+
+impl Backend {
+    pub fn new(client: Client) -> Self {
+        let (rx, tx) = tokio::sync::watch::channel(None);
+
+        Self {
+            client,
+            repo_root: Arc::new(Mutex::new(None)),
+            files: Mutex::new(HashMap::new()),
+            initializer: rx,
+            daemon: tx,
+        }
+    }
+
+    pub async fn package_discovery(&self) -> Result<DiscoveryResponse, discovery::Error> {
+        let mut daemon = {
+            let mut daemon = self.daemon.clone();
+            let daemon = daemon.wait_for(|d| d.is_some()).await;
+            let daemon = daemon.as_ref().expect("only fails if self is dropped");
+            daemon
+                .as_ref()
+                .expect("guaranteed to be some above")
+                .clone()
+        };
+
+        DaemonPackageDiscovery::new(&mut daemon)
+            .discover_packages()
+            .await
+    }
+
+    /// Handle a file update to a rope, emitting diagnostics if necessary.
+    async fn handle_file_update(&self, uri: Url, rope: Option<crop::Rope>, version: Option<i32>) {
+        let rope = match rope {
+            Some(rope) => rope,
+            None => match self.files.lock().expect("only fails if poisoned").get(&uri) {
+                Some(files) => files,
+                None => return,
+            }
+            .clone(),
+        };
+
+        let contents = rope.chunks().join("");
+
+        let repo_root = self
+            .repo_root
+            .lock()
+            .expect("only fails if poisoned")
+            .clone();
+
+        let repo_root = match repo_root {
+            Some(repo_root) => repo_root,
+            None => {
+                self.client
+                    .log_message(MessageType::INFO, "received request before initialization")
+                    .await;
+                return;
+            }
+        };
+
+        let packages = self.package_discovery().await;
+
+        let tasks = packages.map(|p| {
```

Comment:
Not for this PR, but in the future we should expose the logic that the task graph uses for deciding what all of the valid task names so we can make sure the LSP and `turbo` don't drift.